sql: race between RemoteProducerMetaToLocalMeta and DistSQLReceiver.Push

I was running `make testrace PKG=./pkg/sql/opt/exec/execbuilder` and got the data race below.

The line of code in RemoteProducerMetaToLocalMeta is `meta.ContentionEvents = v.ContentionEvents.ContentionEvents`

```
E201204 00:17:21.708869 231374 sql/stats/automatic_stats.go:300  failed to refresh stats: node unavailable; try another peer
==================
WARNING: DATA RACE
Write at 0x00c004d01d68 by goroutine 764:
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.RemoteProducerMetaToLocalMeta()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:271 +0x5ca
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*StreamDecoder).AddMessage()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/stream_decoder.go:103 +0x1a5
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.processProducerMessage()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/inbound.go:193 +0x7c
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.processInboundStreamHelper.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/inbound.go:150 +0x14d

Previous read at 0x00c004d01d68 by goroutine 718:
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:678 +0x124
  github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*copyingRowReceiver).Push()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:439 +0xc1
  github.com/cockroachdb/cockroach/pkg/sql/rowexec.emitHelper()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/processors.go:63 +0x101
  github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sampleAggregator).mainLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sample_aggregator.go:267 +0xd64
  github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sampleAggregator).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sample_aggregator.go:187 +0x12d
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392 +0x43a
  github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Run()
      <autogenerated>:1 +0x84
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:384 +0x787
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunCreateStats()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_stats.go:289 +0x304
  github.com/cockroachdb/cockroach/pkg/sql.(*createStatsResumer).Resume.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:544 +0x284
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5e
  github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xeb
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x124
  github.com/cockroachdb/cockroach/pkg/sql.(*createStatsResumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:532 +0x504
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1136 +0x14c
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1137 +0xa92
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:254 +0x45e
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2cd
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x169

Goroutine 764 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.processInboundStreamHelper()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/inbound.go:132 +0x24b
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.processInboundStream()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/inbound.go:79 +0x9a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.RowInboundStreamHandler.Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/inbound.go:55 +0xc4
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*RowInboundStreamHandler).Run()
      <autogenerated>:1 +0x2e
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).flowStreamInt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:595 +0x5fb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).FlowStream()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:601 +0xa7
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_FlowStream_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:511 +0xcf
  github.com/cockroachdb/cockroach/pkg/util/tracing.StreamServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:168 +0x885
  google.golang.org/grpc.getChainStreamHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1204 +0x123
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:185 +0x108
  google.golang.org/grpc.getChainStreamHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1204 +0x123
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:86 +0xe8
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:63 +0x74
  google.golang.org/grpc.chainStreamServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1190 +0xf9
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1336 +0xbc1
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1409 +0x131e
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xdc

Goroutine 718 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x138
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:852 +0x4c4
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).CreateAndStartJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:301 +0x130
  github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:162 +0x29c
  github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:118 +0x98
==================
FAIL	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder	212.877s
FAIL
make: *** [Makefile:1043: testrace] Error 1
```

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -617,6 +617,9 @@ func (r *DistSQLReceiver) Push(
 	row rowenc.EncDatumRow, meta *execinfrapb.ProducerMetadata,
 ) execinfra.ConsumerStatus {
 	if meta != nil {
+		if metaWriter, ok := r.resultWriter.(MetadataResultWriter); ok {
+			metaWriter.AddMeta(r.ctx, meta)
+		}
 		if meta.LeafTxnFinalState != nil {
 			if r.txn != nil {
 				if r.txn.ID() == meta.LeafTxnFinalState.Txn.ID {
@@ -673,17 +676,15 @@ func (r *DistSQLReceiver) Push(
 				atomic.StoreUint64(r.progressAtomic, math.Float64bits(progress))
 			}
 			meta.Metrics.Release()
-			meta.Release()
 		}
-		if meta.ContentionEvents != nil && r.contendedQueryMetric != nil {
+		if r.contendedQueryMetric != nil && len(meta.ContentionEvents) > 0 {
 			// Increment the contended query metric at most once if the query sees at
 			// least one contention event.
 			r.contendedQueryMetric.Inc(1)
 			r.contendedQueryMetric = nil
 		}
-		if metaWriter, ok := r.resultWriter.(MetadataResultWriter); ok {
-			metaWriter.AddMeta(r.ctx, meta)
-		}
+		// Release the meta object. It is unsafe for use after this call.
+		meta.Release()
 		return r.status
 	}
 	if r.resultWriter.Err() == nil && r.ctx.Err() != nil {

